cpp/cmake_modules/SetupCxxFlags.cmake

```diff
@@ -115,6 +115,9 @@ elseif(ARROW_CPU_FLAG STREQUAL "aarch64")
   if(ARROW_SIMD_LEVEL STREQUAL "DEFAULT")
     set(ARROW_SIMD_LEVEL "NEON")
   endif()
+  if(APPLE)
+    set(GTest_SOURCE "BUNDLED")
+  endif()
```

Comment:
We should not change `*_SOURCE` in `SetupCxxFlags.cmake`.
We should process `*_SOURCE` in `ThirdpartyToolchain.cmake`.

This problem is caused by C++ standard difference, right?
For example, this problem is caused when Apache Arrow C++ uses C++17 and system GoogleTest uses C++14.

Can we detect whether system GoogleTest uses different C++ standard than Apache Arrow C++? If we can detect it, we can fallback to bundled GoogleTest automatically when system GoogleTest uses different C++ standard.

Reply:
Hi Sutou, Good point. I will check CMake and get back to you.

Reply:
After checking the upstream [FindGTest.cmake](https://gitlab.kitware.com/cmake/cmake/blob/master/Modules/FindGTest.cmake) as well as the pkgconfig from GoogleTest on my host, here is my finding:

In order to detect what C++ standard is used by GoogleTest, there is no simple way to do so except by importing it first to my best knowledge. Moreover, once a target is imported, you can not rename or delete it(See Ben's - one of the main CMake developers - comment [here](https://discourse.cmake.org/t/how-to-delete-or-rename-an-imported-target-using-cmake/2372/2?u=haocheng_liu1)). Thus there is no way to fall back to bundled mode.

My new proposed fix is at CMake configuration time, the CMake code will automatically detect and
compare the C++ standard used by GTest and arrow. If a mismatch is detected,
CMake will error out and provide meaningful guidance.  With respect to The principle of least astonishment, I prefer to let the users add these flags by themselves.

```
// sample output from my host
CMake Error at cmake_modules/ThirdpartyToolchain.cmake:2238 (message):
  SYSTEM GTest is built with c++14 but project is built

  with a different standard(17).  Use bundled GTest via passing in CMake flag

  -DGTest_SOURCE="BUNDLED"
Call Stack (most recent call first):
  CMakeLists.txt:505 (include)


-- Configuring incomplete, errors occurred!
```

Looking forward to your thoughts. @kou .

Reply:
Can we use `try_compile()` https://cmake.org/cmake/help/latest/command/try_compile.html for this like we did for gRPC https://github.com/apache/arrow/blob/ddd0a337174e57cdc80b1ee30dc7e787acfc09f6/cpp/src/arrow/flight/CMakeLists.txt#L119-L123 ?

I think that checking `INTERFACE_COMPILE_FEATURES` isn't so portable.
For example, `libgtest-dev` on Debian/Ubuntu still uses `cxx_std_11` not `cxx_std_14`.

For CMake target conflict, we can use different CMake target names (such as `arrow::GTest::gtest`) and CMake variables to refer suitable CMake targets (such as `set(ARROW_GTEST_GTEST arrow::GTest::gtest)`) like we did for Zstandard: https://github.com/apache/arrow/blob/ddd0a337174e57cdc80b1ee30dc7e787acfc09f6/cpp/cmake_modules/ThirdpartyToolchain.cmake#L2577-L2589

Reply:
We can use `FindXXXAlt.cmake` such as `FindGTestAlt.cmake` to check such additional check.
We can use `FindXXXAlt.cmake` by `resolve_dependency(... HAVE_ALT TRUE)`.

Reply:
Also there is no need to do both try_compile and FindXXXAlt.cmake right? The later one sounds an over engineering solution to me….. it will bloat the cmake code in arrow with limited benefits.

Reply:
> - **I think that checking INTERFACE_COMPILE_FEATURES isn't so portable.
For example, libgtest-dev on Debian/Ubuntu still uses cxx_std_11 not cxx_std_14.**

I agree. As [Arow's minimum CXX_STANDARD requirement is 17](https://github.com/apache/arrow/blob/main/cpp/cmake_modules/SetupCxxFlags.cmake#L128-L130), how about We error out whenever `INTERFACE_COMPILE_FEATURES` from GTest is lower than that?  It will be universal to all platforms.
Ty.

Reply:
(PR updated with item 1) )

Reply:
The latter.
We can just wrap existing `GTestConfig.cmake` like we did in `FindOpenSSLAlt.cmake`: https://github.com/apache/arrow/blob/main/cpp/cmake_modules/FindOpenSSLAlt.cmake

> In this PR, We limit the scope to error out whenever `INTERFACE_COMPILE_FEATURES` from GTest is lower than that? It will be universal to all platforms and truly detect the issue at CMake configure time.
> 
> Then in another PR/issue, we can discuss how to implement try_compile and FindXXXAlt.cmake as it will need non trivial work.

It makes sense. Could you update the title of this pull request and open a new issue for the follow-up action?

Reply:
FYI: If we use the latter approach, we can implement an automatically fallback solution instead of raising an error for C++ standard miss match. (If system GoogleTest isn't suitable, we can use our bundled GoogleTest instead of stopping `cmake`.)

Reply:
Thanks for the suggestion @kou , I've updated the title of the PR and opened a new follow up issue. Please take a look when you have time

Reply:
Thanks!

Reply:
[The failing test ](https://github.com/apache/arrow/actions/runs/4570445331/jobs/8068019026?pr=34765)- test_generic_options - looks like a flaky one